lib/container.ts

```diff
@@ -207,55 +207,127 @@ export class Container extends (EventEmitter as {
 	}
 
-	private getAddOperations(
+	private async getAddOperations(
 		files: string[],
 		actionGroup: DockerfileActionGroup,
-	): AddOperation[] {
-		return this.getOperations(files, actionGroup);
+	): Promise<AddOperation[]> {
+		return await this.getOperations(files, actionGroup);
 	}
 
-	private getDeleteOperations(
+	private async getDeleteOperations(
 		files: string[],
 		actionGroup: DockerfileActionGroup,
-	): DeleteOperation[] {
+	): Promise<DeleteOperation[]> {
 		// for every file that we need to delete, find out where it was sent to
 		// and return the in-container path
-		return _.map(this.getOperations(files, actionGroup), 'toPath');
+		return _.map(await this.getOperations(files, actionGroup), 'toPath');
 	}
 
-	private getOperations(
+	private async getOperations(
 		files: string[],
 		actionGroup: DockerfileActionGroup,
-	): Array<{ fromPath: string; toPath: string }> {
-		return files.map(f => {
-			const matchingDep = Dockerfile.getActionGroupFileDependency(
-				f,
-				actionGroup,
-			);
-
-			/* istanbul ignore next */
-			if (matchingDep == null) {
-				throw new InternalInconsistencyError(
-					'Could not find matching file for action group',
+	): Promise<Array<{ fromPath: string; toPath: string }>> {
+		return Promise.all(
+			files.map(async f => {
+				const matchingDep = Dockerfile.getActionGroupFileDependency(
+					f,
+					actionGroup,
 				);
-			}
 
-			// TODO: I'm not sure the logic here is actually correct,
-			// specifically the way we build the destination when the containerPath
-			// is a directory
-			const strippedPath = matchingDep.sourceIsDirectory
-				? f
-				: f.split(Path.sep).pop()!;
-
-			const toPath = matchingDep.destinationIsDirectory
-				? Path.join(matchingDep.containerPath, strippedPath)
-				: matchingDep.containerPath;
-
-			return {
-				fromPath: f,
-				toPath,
-			};
+				/* istanbul ignore next */
+				if (matchingDep == null) {
+					throw new InternalInconsistencyError(
+						'Could not find matching file for action group',
+					);
+				}
+
+				// TODO: I'm not sure the logic here is actually correct,
+				// specifically the way we build the destination when the containerPath
+				// is a directory
+
+				// The dockerfile class tries to detect if the destination is a directory,
+				// but we can do one better with the actual container, and look directly
+				// at the destination
+				if (!matchingDep.destinationIsDirectory) {
+					matchingDep.destinationIsDirectory = await this.containerPathIsDirectory(
+						matchingDep.containerPath,
+					);
+				}
+
+				// We can also check if the host path is a directory (only if it's not obvious
+				// from the dockerfile)
+				if (!matchingDep.sourceIsDirectory) {
+					matchingDep.sourceIsDirectory = await Container.hostPathIsDirectory(
+						Path.join(this.hostContextPath, f),
+					);
+				}
+
+				const strippedPath = matchingDep.sourceIsDirectory
+					? f
+					: f.split(Path.sep).pop()!;
```

Comment:
I think this would be better as
```suggestion
					: Path.basename(f);
```